### ISSUE TITLE
IllegalStateException whilst planning distinct union

### ISSUE THREAD

**Issue by InverseIntegral** [neo4j/neo4j#12911]
The following query
```
MATCH (n:L) WHERE NOT(((NOT(toBoolean(true))) AND NOT(isEmpty(n.p))) XOR (isEmpty(n.p))) RETURN *
```
throws an unexpected exception:
```
java.lang.IllegalStateException: Planning a distinct for union, but no union was planned before.
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.LogicalPlanProducer.planDistinctForUnion(LogicalPlanProducer.scala:1365)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.OrLeafPlanner.mergePlansWithUnion$1(OrLeafPlanner.scala:377)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.OrLeafPlanner.$anonfun$apply$30(OrLeafPlanner.scala:397)
	at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:935)
	at scala.collection.IndexedSeqOptimized.foreach(IndexedSeqOptimized.scala:36)
	at scala.collection.IndexedSeqOptimized.foreach$(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:198)
	at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:934)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.OrLeafPlanner.$anonfun$apply$28(OrLeafPlanner.scala:395)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.OrLeafPlanner.$anonfun$apply$28$adapted(OrLeafPlanner.scala:386)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.OrLeafPlanner.$anonfun$apply$22(OrLeafPlanner.scala:386)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.immutable.Set$Set2.foreach(Set.scala:181)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.OrLeafPlanner.apply(OrLeafPlanner.scala:385)
	at org.neo4j.cypher.internal.compiler.planner.logical.LeafPlannerList.$anonfun$candidates$1(LeafPlannerIterable.scala:43)
	at scala.collection.TraversableLike.$anonfun$flatMap$1(TraversableLike.scala:293)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.flatMap(TraversableLike.scala:293)
	at scala.collection.TraversableLike.flatMap$(TraversableLike.scala:290)
	at scala.collection.AbstractTraversable.flatMap(Traversable.scala:108)
	at org.neo4j.cypher.internal.compiler.planner.logical.LeafPlannerList.candidates(LeafPlannerIterable.scala:43)
	at org.neo4j.cypher.internal.compiler.planner.logical.steps.leafPlanOptions$.apply(leafPlanOptions.scala:58)
	at org.neo4j.cypher.internal.compiler.planner.logical.idp.SingleComponentPlanner.planComponent(SingleComponentPlanner.scala:53)
	at org.neo4j.cypher.internal.compiler.planner.logical.idp.IDPQueryGraphSolver.$anonfun$planComponents$1(IDPQueryGraphSolver.scala:131)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:286)
	at scala.collection.Iterator.foreach(Iterator.scala:943)
	at scala.collection.Iterator.foreach$(Iterator.scala:943)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1431)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.map(TraversableLike.scala:286)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:279)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at org.neo4j.cypher.internal.compiler.planner.logical.idp.IDPQueryGraphSolver.planComponents(IDPQueryGraphSolver.scala:130)
	at org.neo4j.cypher.internal.compiler.planner.logical.idp.IDPQueryGraphSolver.plan(IDPQueryGraphSolver.scala:110)
	at org.neo4j.cypher.internal.compiler.planner.logical.planMatch$.doPlan(planMatch.scala:38)
	at org.neo4j.cypher.internal.compiler.planner.logical.MatchPlanner.plan(PlanSingleQuery.scala:104)
	at org.neo4j.cypher.internal.compiler.planner.logical.MatchPlanner.plan$(PlanSingleQuery.scala:103)
	at org.neo4j.cypher.internal.compiler.planner.logical.planMatch$.plan(planMatch.scala:26)
	at org.neo4j.cypher.internal.compiler.planner.logical.PlanHead.plan(PlanHead.scala:47)
	at org.neo4j.cypher.internal.compiler.planner.logical.PlanSingleQuery.plan(PlanSingleQuery.scala:49)
	at org.neo4j.cypher.internal.compiler.planner.logical.plannerQueryPartPlanner$.plan(QueryPlanner.scala:162)
	at org.neo4j.cypher.internal.compiler.planner.logical.QueryPlanner$.plan(QueryPlanner.scala:114)
	at org.neo4j.cypher.internal.compiler.planner.logical.QueryPlanner$.process(QueryPlanner.scala:63)
	at org.neo4j.cypher.internal.compiler.planner.logical.QueryPlanner$.process(QueryPlanner.scala:51)
	at org.neo4j.cypher.internal.frontend.phases.Phase.$anonfun$transform$1(Phase.scala:36)
	at org.neo4j.cypher.internal.frontend.helpers.package$.$anonfun$closing$1(package.scala:24)
	at org.neo4j.cypher.internal.frontend.helpers.package$.using(package.scala:32)
	at org.neo4j.cypher.internal.frontend.helpers.package$.closing(package.scala:24)
	at org.neo4j.cypher.internal.frontend.phases.Phase.transform(Phase.scala:35)
	at org.neo4j.cypher.internal.frontend.phases.Phase.transform$(Phase.scala:33)
	at org.neo4j.cypher.internal.compiler.planner.logical.QueryPlanner$.transform(QueryPlanner.scala:51)
	at org.neo4j.cypher.internal.frontend.phases.PipeLine.transform(Transformer.scala:79)
	at org.neo4j.cypher.internal.frontend.phases.PipeLine.transform(Transformer.scala:78)
	at org.neo4j.cypher.internal.frontend.phases.PipeLine.transform(Transformer.scala:78)
	at org.neo4j.cypher.internal.frontend.phases.PipeLine.transform(Transformer.scala:78)
	at org.neo4j.cypher.internal.frontend.phases.PipeLine.transform(Transformer.scala:78)
	at org.neo4j.cypher.internal.frontend.phases.If.transform(Transformer.scala:96)
	at org.neo4j.cypher.internal.frontend.phases.PipeLine.transform(Transformer.scala:79)
	at org.neo4j.cypher.internal.compiler.CypherPlanner.planPreparedQuery(CypherPlanner.scala:70)
	at org.neo4j.cypher.internal.planning.CypherPlanner.doCreatePlan(CypherPlanner.scala:404)
	at org.neo4j.cypher.internal.planning.CypherPlanner.org$neo4j$cypher$internal$planning$CypherPlanner$$createPlan$1(CypherPlanner.scala:352)
	at org.neo4j.cypher.internal.planning.CypherPlanner$$anon$1.compile(CypherPlanner.scala:361)
	at org.neo4j.cypher.internal.planning.CypherPlanner$$anon$1.compile(CypherPlanner.scala:360)
	at org.neo4j.cypher.internal.QueryCache.$anonfun$compileAndCache$1(QueryCache.scala:276)
	at org.neo4j.cypher.internal.QueryCache.compileOrCompileWithExpressionCodeGenAndCache(QueryCache.scala:304)
	at org.neo4j.cypher.internal.QueryCache.compileAndCache(QueryCache.scala:276)
	at org.neo4j.cypher.internal.QueryCache.computeIfAbsentOrStale(QueryCache.scala:193)
	at org.neo4j.cypher.internal.planning.CypherPlanner.doPlan(CypherPlanner.scala:379)
	at org.neo4j.cypher.internal.planning.CypherPlanner.parseAndPlan(CypherPlanner.scala:269)
	at org.neo4j.cypher.internal.CypherCurrentCompiler.compile(CypherCurrentCompiler.scala:152)
	at org.neo4j.cypher.internal.MasterCompiler.compile(MasterCompiler.scala:86)
	at org.neo4j.cypher.internal.ExecutionEngine.$anonfun$compilerWithExpressionCodeGenOption$2(ExecutionEngine.scala:229)
	at org.neo4j.cypher.internal.ExecutionEngine$$anon$2.compile(ExecutionEngine.scala:237)
	at org.neo4j.cypher.internal.ExecutionEngine$$anon$2.compile(ExecutionEngine.scala:231)
	at org.neo4j.cypher.internal.QueryCache.$anonfun$compileAndCache$1(QueryCache.scala:276)
	at org.neo4j.cypher.internal.QueryCache.compileOrCompileWithExpressionCodeGenAndCache(QueryCache.scala:304)
	at org.neo4j.cypher.internal.QueryCache.compileAndCache(QueryCache.scala:276)
	at org.neo4j.cypher.internal.QueryCache.computeIfAbsentOrStale(QueryCache.scala:193)
	at org.neo4j.cypher.internal.ExecutionEngine.getOrCompile(ExecutionEngine.scala:287)
	at org.neo4j.cypher.internal.ExecutionEngine.doExecute(ExecutionEngine.scala:200)
	at org.neo4j.cypher.internal.ExecutionEngine.$anonfun$executeSubquery$1(ExecutionEngine.scala:178)
	at org.neo4j.cypher.internal.ExecutionEngine.closing(ExecutionEngine.scala:183)
	at org.neo4j.cypher.internal.ExecutionEngine.executeSubquery(ExecutionEngine.scala:175)
	at org.neo4j.cypher.internal.ExecutionEngine.execute(ExecutionEngine.scala:125)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:127)
	at org.neo4j.cypher.internal.javacompat.ExecutionEngine.executeQuery(ExecutionEngine.java:115)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:285)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:274)
	at org.neo4j.kernel.impl.coreapi.TransactionImpl.execute(TransactionImpl.java:268)
```
I would expect the query to run without any exception.

**Neo4j Version:** 4.4.9
**Operating System:** 5.18.9 Arch Linux
**API:** Java

**Comment by HannesSandberg** [comment]
Thank you for the report! We will investigate and come back to you.

**Comment by sherfert** [comment]
@InverseIntegral The fix will be part of the next 4.4 release.
